File: backend/billing.py
"""
Monetization: charge for quiet, never for dignity (POPPY_PRODUCT_PLAYBOOK §8).

**One paid thing, and it is not a feature.** Free is the whole product: every call,
every vibe, the full memory, the rituals, unlimited. Plus is the same product with the
ads switched off, bought once. There is no tier that talks to her more, because a
companion that meters conversation is a companion that fails the person who needs the
sixth call more than the first.

That leaves exactly one guardrail worth enforcing in code, and it is the one this file
has always had: **an interruption may only land at an abundance moment.** It used to
mean a paywall. It now means an ad, which needs it more — an ad on top of someone
venting is worse than a paywall there, and is the fastest way to lose the trust the
whole product runs on. `can_interrupt()` makes both impossible, the same way nudges.py
makes a guilt-trip impossible.

The entitlement is a StoreKit / Play Billing non-consumable. Unlike the credit ledger in
accounts.py, it is genuinely enforceable without a server: `Transaction.currentEntitlements`
and `queryPurchasesAsync` are signed by the store and verified on-device. On desktop it
stays a local field on the profile.
"""

import companion

# Two tiers, and the difference between them is one word long.
#
# The copy matters more than usual here: whatever is written in `features` is what the
# store listing promises. The old text promised Plus "unlimited, longer calls" back when
# Free was capped at five a day. Free is uncapped now, so that line would be a false
# claim on two storefronts, not merely stale marketing.
TIERS = {
    "free": {
        "name": "Free",
        "price": "₹0",
        "blurb": "The whole thing, free.",
        "features": [
            "Unlimited calls, always",
            "Your companion, your vibe",
            "Core memory that remembers you",
            "Morning / night ritual",
            "Full privacy, export, and delete",
            "A few ads between conversations",
        ],
    },
    "plus": {
        "name": "Poppy Plus",
        "price": "$20 once",
        "blurb": "The same Poppy, without the ads.",
        "features": [
            "No ads, anywhere, ever",
            "Everything in Free, unchanged",
            "One payment, not a subscription",
        ],
    },
}

# Mood modes that are emotionally vulnerable by nature. Nothing interrupts these.
_VULNERABLE_MODES = {"vent", "wind"}

# Two switches, because the two halves ship at different times.
#
# ADS_LIVE says ads may be requested. BILLING_LIVE says the upgrade may be offered.
# They were one flag and that was wrong: it forced the sell and the buy to arrive
# together, when the useful order is ads first (with Google's test units, which need no
# account and cost nothing to get wrong) and the purchase after.
#
# The dangerous combination is BILLING_LIVE without a store bridge, because the button
# would take no money and grant Plus anyway. So that one stays False until StoreKit and
# Play Billing are wired. Ads with no purchase are merely annoying, not broken, and are
# the thing being tested right now.
ADS_LIVE = True
BILLING_LIVE = False


def plan() -> str:
    return companion.profile().get("plan", "free")


def can_interrupt(context: dict | None = None) -> bool:
    """The §8 guardrail, and the only gate in this file.

    False for any vulnerable moment: a distress/crisis-flagged turn, or an emotionally
    vulnerable mood mode. Every ad surface and every upgrade prompt must pass through
    here, so neither can appear at a moment where it would cost more than it earns.
    """
    ctx = context or {}
    if ctx.get("crisis") or ctx.get("distress"):
        return False
    if ctx.get("mode") in _VULNERABLE_MODES:
        return False
    return True


def should_show_ads(context: dict | None = None) -> bool:
    """True when an ad may be requested right now: the user has not bought Plus, and
    this is an abundance moment. Check this at *every* ad call site — a paid user must
    never see a request fire, not even one that fails to fill."""
    if not ADS_LIVE:
        return False
    if plan() != "free":
        return False
    return can_interrupt(context)


def entitlement() -> dict:
    """Current tier and, the only thing any caller actually branches on, whether ads
    are on. No counters: there is nothing left to count."""
    p = plan()
    return {
        "plan": p,
        "billing_live": BILLING_LIVE,
        "ads": ADS_LIVE and p == "free",
        "tier": TIERS.get(p, TIERS["free"]),
        "tiers": TIERS,
    }


def set_plan(new_plan: str) -> dict:
    """Record the tier. The *authority* for this is the store receipt, not this call:
    mobile resolves the entitlement from StoreKit / Play Billing on every cold start
    and calls this to cache the result. Desktop has no store, so here it is the truth."""
    new_plan = new_plan if new_plan in TIERS else "free"
    companion.update(plan=new_plan)
    return entitlement()


def referral() -> dict:
    """A share code for the aligned-incentive referral loop (§7 loop B). Local stub on
    desktop; real redemption/attribution is a thin-cloud job (D2)."""
    import uuid
    p = companion.profile()
    code = p.get("referral_code")
    if not code:
        code = "POPPY-" + uuid.uuid4().hex[:6].upper()
        companion.update(referral_code=code)
    return {
        "code": code,
        "message": "Give a friend Poppy, ad-free for a week, and get a week yourself.",
    }

File: mobile/src/core/billing.ts
/**
 * Plans and the ad guardrail — the port of backend/billing.py. Keep the two in step.
 *
 * ## What changed, and why the default flipped
 *
 * `DEFAULT_PLAN` was `plus`, because there was no price and no StoreKit product, and a
 * paywall with no way to pay is a door with no handle. There is a price now, and the
 * thing being sold is ads-off, so the same reasoning inverts: leaving everyone on
 * `plus` would mean no phone ever shows an ad and the free tier does not exist.
 * It is `free`, and StoreKit / Play Billing move a user off it.
 *
 * ## What did not change, and is the whole point of this file
 *
 * Free is the entire product, uncapped. The daily-call limit is gone: metering
 * conversation is the one thing this app will not do, and the person reaching for a
 * sixth call today is the one least well served by a locked door. $20 buys quiet.
 *
 * So one rule survives, and it matters more now than it ever did as a paywall rule:
 * **nothing interrupts a vulnerable moment.** An ad on top of someone venting is worse
 * than a paywall there. `canInterrupt()` makes it impossible rather than discouraged.
 */

import * as companion from './companion';

export const TIERS: Record<string, {
  name: string;
  price: string;
  blurb: string;
  features: string[];
}> = {
  free: {
    name: 'Free',
    price: '₹0',
    blurb: 'The whole thing, free.',
    features: [
      'Unlimited calls, always',
      'Your companion, your vibe',
      'Core memory that remembers you',
      'Morning / night ritual',
      'Full privacy, export, and delete',
      'A few ads between conversations',
    ],
  },
  plus: {
    name: 'Poppy Plus',
    price: '$20 once',
    blurb: 'The same Poppy, without the ads.',
    features: [
      'No ads, anywhere, ever',
      'Everything in Free, unchanged',
      'One payment, not a subscription',
    ],
  },
};

/** Modes where nothing is ever sold and nothing ever interrupts. */
const VULNERABLE_MODES = new Set(['vent', 'wind']);

/** There is a price now, so the free tier is real and is where everyone starts. */
const DEFAULT_PLAN = 'free';

/**
 * Two switches, because the two halves ship at different times. Keep in step with
 * billing.py.
 *
 * `ADS_LIVE` says ads may be requested. `BILLING_LIVE` says the upgrade may be offered.
 * They were one flag and that was wrong: it forced the sell and the buy to arrive
 * together, when the useful order is ads first (on Google's test units, which need no
 * account and cost nothing to get wrong) and the purchase after.
 *
 * The dangerous combination is `BILLING_LIVE` with no store bridge, because the button
 * would take no money and grant Plus anyway. That stays false until StoreKit and Play
 * Billing are wired. Ads without a purchase are merely annoying, not broken.
 */
export const ADS_LIVE = true;
const BILLING_LIVE = false;

export async function plan(): Promise<string> {
  // Nobody can have bought Plus while billing is not live, so a saved 'plus' cannot be
  // a purchase. It is the old default: companion.ts shipped `plan: 'plus'` back when
  // there was no price, and update() writes the whole profile, so every phone that ever
  // ran that build has it persisted. Trusting it would mean no existing install ever
  // sees an ad. Once billing is live the store receipt is re-resolved on every cold
  // start and overwrites this field, so the saved value becomes meaningful again.
  if (!BILLING_LIVE) return 'free';
  const p = await companion.profile();
  const saved = p.plan;
  return saved && TIERS[saved] ? saved : DEFAULT_PLAN;
}

/**
 * The guardrail. False for any vulnerable moment — a distress or crisis turn, or an
 * emotionally vulnerable mood. Every ad surface and every upgrade prompt goes through
 * here.
 */
export function canInterrupt(
  context: { crisis?: boolean; distress?: boolean; mode?: string } = {},
): boolean {
  if (context.crisis || context.distress) return false;
  if (context.mode && VULNERABLE_MODES.has(context.mode)) return false;
  return true;
}

/**
 * True when an ad may be requested right now. Check at *every* ad call site: a paid
 * user must never see a request fire, not even one that fails to fill.
 */
export async function shouldShowAds(
  context: { crisis?: boolean; distress?: boolean; mode?: string } = {},
): Promise<boolean> {
  if (!ADS_LIVE) return false;
  if ((await plan()) !== 'free') return false;
  return canInterrupt(context);
}

export async function entitlement(): Promise<Record<string, unknown>> {
  const p = await plan();
  return {
    plan: p,
    billing_live: BILLING_LIVE,
    // The only thing any caller branches on.
    ads: ADS_LIVE && p === 'free',
    tier: TIERS[p] ?? TIERS.free,
    tiers: TIERS,
  };
}

/**
 * Cache the tier. The authority is the store receipt, not this call: StoreKit's
 * `Transaction.currentEntitlements` and Play's `queryPurchasesAsync` are signed and
 * verified on-device, and are re-resolved on every cold start. This only records what
 * they said, so a wrong value here is corrected by the next launch rather than sold.
 */
export async function setPlan(next: string): Promise<Record<string, unknown>> {
  await companion.update({ plan: TIERS[next] ? next : 'free' });
  return entitlement();
}
